### PR TITLE
GRPC OnlineQueryBulk 

### PR DESCRIPTION
**Pull request by rooftoofwoof**
Adds a GRPC client and implement the online query bulk method

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -194,6 +201,19 @@ type ClientConfig struct {
...
 //
 // [chalk login]: https://docs.chalk.ai/cli#login
-func NewClient(config ...*ClientConfig) (Client, error) {
-	return newClientImpl(config...)
+func NewClient(configs ...*ClientConfig) (Client, error) {
+	var cfg *ClientConfig
+	if len(configs) == 0 {
+		cfg = &ClientConfig{}
+	} else if len(configs) > 1 {
+		return nil, fmt.Errorf("expected at most one ClientConfig, got %d", len(configs))
```
Comment (rooftoofwoof) [inline client.go:209]: errors.Newf

File: client_grpc.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+func (c *clientGrpc) NewAuthClient() (serverv1connect.AuthServiceClient, error) {
+	return serverv1connect.NewAuthServiceClient(
+		c.httpClient,
+		c.config.apiServer.Value,
+		withChalkInterceptors(
+			serverTypeApi,
+			headerInterceptor(map[string]string{
+				headerKeyServerType: "go-api",
```
Comment (rooftoofwoof) [inline client_grpc.go:101]: use constant

File: client_grpc.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+		return nil, errors.Newf(
+			"no engine found for environment '%s' - engine map keys: '%s'",
+			c.config.environmentId.Value,
+			lo.Keys(c.config.engines),
+		)
+	}
+
+	return enginev1connect.NewQueryServiceClient(
+		c.httpClient,
+		"https://"+endpoint,
```
Comment (rooftoofwoof) [inline client_grpc.go:123]: fmt sprintf? or use a url lib?

File: internal/auth/utils.go [hunk 4]
```diff
@@ -143,3 +143,10 @@ type SourcedConfig struct {
 	Value  string
 	Source string
 }
+
+type ResolvedConfig struct {
```
Comment (rooftoofwoof) [inline internal/auth/utils.go:147]: remove

File: go.mod [hunk 5]
```diff
@@ -14,17 +15,29 @@ require (
...
 	github.com/klauspost/cpuid/v2 v2.2.8 // indirect
+	github.com/kr/pretty v0.3.1 // indirect
+	github.com/kr/text v0.2.0 // indirect
+	github.com/mattn/go-colorable v0.1.13 // indirect
+	github.com/mattn/go-isatty v0.0.19 // indirect
 	github.com/pierrec/lz4/v4 v4.1.21 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/rogpeppe/go-internal v1.9.0 // indirect
+	github.com/rs/zerolog v1.33.0 // indirect
+	github.com/samber/lo v1.46.0 // indirect
```
Comment (rooftoofwoof) [inline go.mod:35]: go mod tidy

File: client_impl.go [hunk 6]
```diff
@@ -474,13 +447,13 @@ func (c *clientImpl) sendRequest(args sendRequestParams) error {
 	request.Header = headers
 
 	if !args.DontRefresh {
-		upsertJwtErr := c.refreshConfig(false)
+		upsertJwtErr := c.config.refresh(false)
 		if upsertJwtErr != nil {
```
Comment (rooftoofwoof) [inline client_impl.go:451]: inline

File: client_impl.go [hunk 7]
```diff
@@ -649,66 +622,34 @@ func getErrorResponse(err error) *ErrorResponse {
...
+		httpClient: httpClient,
 
-	err := client.refreshConfig(false)
-	if cfg.Logger != nil {
-		client.logger = cfg.Logger
+		config: config,
 	}
+	client.config.getToken = client.getToken
+	err = client.config.refresh(false)
 	if err != nil {
```
Comment (rooftoofwoof) [inline client_impl.go:653]: inline

File: from_proto.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+		DeploymentId:       m.DeploymentId,
+		EnvironmentId:      m.EnvironmentId,
+		EnvironmentName:    m.EnvironmentName,
+		QueryId:            m.QueryId,
+		QueryTimestamp:     lo.ToPtr(m.QueryTimestamp.AsTime()),
+		QueryHash:          m.QueryHash,
+	}
+}
+
+var errorCodeMap = map[commonv1.ErrorCode]ErrorCode{
```
Comment (rooftoofwoof) [inline from_proto.go:24]: check if exhaustive

File: internal/auth/utils.go (hunk 4 above)
Comment (rooftoofwoof) [inline internal/auth/utils.go:147]: Remove

File: client_grpc.go [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+			GrantType:    "client_credentials",
+		},
+	)
+	token, err := c.authClient.GetToken(context.Background(), authRequest)
+	if err != nil {
+		c.logger.Debugf("Failed to get a new token: %s", err.Error())
+		return nil, err
+	}
+	return &getTokenResult{
+		ValidUntil:         token.Msg.GetExpiresAt().AsTime(),
```
Comment (rooftoofwoof) [inline client_grpc.go:170]: expiresat might be nil

**Review by emarx — APPROVED** [review]

**Pull request merged**